### PR TITLE
make sure the library builds with C++17 and C++20, in addition to C++14

### PR DESCRIPTION
**Pull request by arvidn**

### CODE REVIEW

**Comment by Chocobo1** [comment]
>rely on RVO in C++20

RC_1_2 branch is also affected, please consider a backport, thanks!

File: include/libtorrent/config.hpp [hunk 1]
```diff
@@ -616,6 +616,12 @@ POSSIBILITY OF SUCH DAMAGE.
 #define __has_builtin(x) 0  // for non-clang compilers
 #endif
 
+#if __cplusplus >= 202002L
```
[the remark this answers is not in the document]
Comment (arvidn) [reply include/libtorrent/config.hpp:619]: it doesn't look like these are issues on `RC_2_0` thgough

[the remark this answers is not in the document]
Comment (arvidn) [reply include/libtorrent/config.hpp:619]: I'm not sure why those aren't called out by the compiler. but I fixed them anyway

[the remark this answers is not in the document]
Comment (Chocobo1) [reply include/libtorrent/config.hpp:619]: FYI, I only noticed them when my compiler emitted warnings.

**Pull request merged**